### Importing Libraries

In [24]:
# Import libraries
import pandas as pd
import numpy as np
import os

### Importing Data

In [26]:
path = r"C:\Users\nabar\OneDrive\Documents\03-2025 Instacart Basket Analysis"
orders_products_combined = pd.read_pickle(os.path.join(path, '02 Data' , 'Prepared Data' , 'orders_products_combined.pkl'))
df_prods = pd.read_csv(os.path.join(path, '02 Data' , 'Prepared Data' , 'products_checked.csv'), index_col = 0)

In [27]:
orders_products_combined.shape

(32434489, 11)

In [28]:
# Check the output
orders_products_combined.head()

order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0   2539329        1    prior             1                       2   
1   2539329        1    prior             1                       2   
2   2539329        1    prior             1                       2   
3   2539329        1    prior             1                       2   
4   2539329        1    prior             1                       2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [29]:
# Check the output
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [30]:
df_prods.shape

(49672, 5)

### Merging Dataframes

Product_id is the key column.

In [33]:
# Test Merge
pd.merge(orders_products_combined , df_prods , on = ['product_id'] , how = 'inner', indicator = 'match' )

order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0          2539329        1    prior             1                       2   
1          2539329        1    prior             1                       2   
2          2539329        1    prior             1                       2   
3          2539329        1    prior             1                       2   
4          2539329        1    prior             1                       2   
...            ...      ...      ...           ...                     ...   
32404854   2977660   206209    prior            13                       1   
32404855   2977660   206209    prior            13                       1   
32404856   2977660   206209    prior            13                       1   
32404857   2977660   206209    prior            13                       1   
32404858   2977660   206209    prior            13                       1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32404854                 12                     7.0       14197   
32404855                 12                     7.0       38730   
32404856                 12                     7.0       31477   
32404857                 12                     7.0        6567   
32404858                 12                     7.0       22920   

          add_to_cart_order  reordered _merge  \
0                         1          0   both   
1                         2          0   both   
2                         3          0   both   
3                         4          0   both   
4                         5          0   both   
...                     ...        ...    ...   
32404854                  5          1   both   
32404855                  6          0   both   
32404856                  7          0   both   
32404857                  8          0   both   
32404858                  9          0   both   

                                     product_name  aisle_id  department_id  \
0                                            Soda        77              7   
1         Organic Unsweetened Vanilla Almond Milk        91             16   
2                             Original Beef Jerky        23             19   
3                      Aged White Cheddar Popcorn        23             19   
4                XL Pick-A-Size Paper Towel Rolls        54             17   
...                                           ...       ...            ...   
32404854                             Tomato Paste         9              9   
32404855          Brownie Crunch High Protein Bar         3             19   
32404856    High Protein Bar Chunky Peanut Butter         3             19   
32404857      Chocolate Peanut Butter Protein Bar         3             19   
32404858      Roasted & Salted Shelled Pistachios       117             19   

          prices match  
0            9.0  both  
1           12.5  both  
2            4.4  both  
3            4.7  both  
4            1.0  both  
...          ...   ...  
32404854     5.6  both  
32404855     5.9  both  
32404856     4.2  both  
32404857     4.9  both  
32404858     1.7  both  

[32404859 rows x 16 columns]

In [34]:
# Merging 'orders_products_combined' and '_df_pords'
ords_prods_merge = orders_products_combined.merge(df_prods , on = ['product_id'] , indicator = 'match' )

In [35]:
# Confirm the results of the merge using the merge flag
ords_prods_merge.drop('_merge', axis = 1 )

order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0          2539329        1    prior             1                       2   
1          2539329        1    prior             1                       2   
2          2539329        1    prior             1                       2   
3          2539329        1    prior             1                       2   
4          2539329        1    prior             1                       2   
...            ...      ...      ...           ...                     ...   
32404854   2977660   206209    prior            13                       1   
32404855   2977660   206209    prior            13                       1   
32404856   2977660   206209    prior            13                       1   
32404857   2977660   206209    prior            13                       1   
32404858   2977660   206209    prior            13                       1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32404854                 12                     7.0       14197   
32404855                 12                     7.0       38730   
32404856                 12                     7.0       31477   
32404857                 12                     7.0        6567   
32404858                 12                     7.0       22920   

          add_to_cart_order  reordered  \
0                         1          0   
1                         2          0   
2                         3          0   
3                         4          0   
4                         5          0   
...                     ...        ...   
32404854                  5          1   
32404855                  6          0   
32404856                  7          0   
32404857                  8          0   
32404858                  9          0   

                                     product_name  aisle_id  department_id  \
0                                            Soda        77              7   
1         Organic Unsweetened Vanilla Almond Milk        91             16   
2                             Original Beef Jerky        23             19   
3                      Aged White Cheddar Popcorn        23             19   
4                XL Pick-A-Size Paper Towel Rolls        54             17   
...                                           ...       ...            ...   
32404854                             Tomato Paste         9              9   
32404855          Brownie Crunch High Protein Bar         3             19   
32404856    High Protein Bar Chunky Peanut Butter         3             19   
32404857      Chocolate Peanut Butter Protein Bar         3             19   
32404858      Roasted & Salted Shelled Pistachios       117             19   

          prices match  
0            9.0  both  
1           12.5  both  
2            4.4  both  
3            4.7  both  
4            1.0  both  
...          ...   ...  
32404854     5.6  both  
32404855     5.9  both  
32404856     4.2  both  
32404857     4.9  both  
32404858     1.7  both  

[32404859 rows x 15 columns]

In [36]:
ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

The merged dataframe contains 32,404,859 rows, where each row includes data from both input datasets. This result is achieved through the use of an inner join, ensuring that only matching records from both datasets are included.

### Exporting Data

In [52]:
ords_prods_merge.drop('_merge', axis = 1, inplace=True )

In [54]:
print(ords_prods_merge.columns)

Index(['order_id', 'user_id', 'eval_set', 'order_number',
       'orders_day_of_the_week', 'order_hour_of_day', 'days_since_prior_order',
       'product_id', 'add_to_cart_order', 'reordered', 'product_name',
       'aisle_id', 'department_id', 'prices', 'match'],
      dtype='object')


In [56]:
# Export data to pkl
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))